File: examples/python/core/steps/exponential-backoff.py
from aws_durable_execution_sdk_python import (
    DurableContext,
    durable_execution,
)
from aws_durable_execution_sdk_python.config import StepConfig
from aws_durable_execution_sdk_python.retries import (
    RetryStrategyConfig,
    create_retry_strategy,
)

@durable_execution
def handler(event: dict, context: DurableContext) -> str:
    # Configure exponential backoff
    retry_config = RetryStrategyConfig(
        max_attempts=3,
        initial_delay_seconds=1,
        max_delay_seconds=10,
        backoff_rate=2.0,
    )
    
    step_config = StepConfig(
        retry_strategy=create_retry_strategy(retry_config)
    )
    
    result = context.step(
        lambda _: "Step with exponential backoff",
        name="retry_step",
        config=step_config,
    )
    return f"Result: {result}"

File: examples/java/core/steps/exponential-backoff.java
public String handleRequest(Map<String, Object> event, DurableContext ctx) {
    // Configure exponential backoff
    var stepConfig = StepConfig.builder()
        .retryStrategy(RetryStrategies.exponentialBackoff(
            3,                        // max attempts
            Duration.ofSeconds(1),    // initial delay
            Duration.ofSeconds(10),   // max delay
            2.0,                      // backoff rate
            JitterStrategy.FULL))
        .build();

    var result = ctx.step("retry-step", String.class, stepCtx -> {
        return "Step with exponential backoff";
    }, stepConfig);

    return "Result: " + result;
}
